proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,190 @@
+An upstreaming package source is one that either aggregates multiple package sources or caches packages from another package source.
+Upstreaming isn't a concept that exists in NuGet, so servers have to present to clients as if all data (packages, search results, vulnerability information) originates from itself.
+
+However, there are also cases where customers are not using upstreaming feeds, but instead curate approved 3rd party packages and manually upload them to an internal feed.
+There are also other scenarios where customers do not use nuget.org (directly) as a package source.
+
+It's possible for customers to use [Package Source Mapping](https://learn.microsoft.com/nuget/consume-packages/package-source-mapping) to add nuget.org as a package source, but configured in a way that NuGet will never restore packages from this package source.
+However, some developers might consider this an unacceptable risk and prefer another way to get a VDB that cannot accidentally be used to get packages.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
+Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />`.
+When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
+
+For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:
+
+```xml
+<configuration>
+  <packageSources>
+    <clear />
+    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
+  </packageSources>
+  <auditSources>
+    <clear />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >
+  </auditSources>
+</configuration>
+```
+
+Any URL specified in the `auditSources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).
+However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info) (and service index), and can avoid implementing all the other resources that are required to be used as a package source.
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+Not using `packageSources` for VDB when there's at least one source in `auditSources` is a performance optimization.
```

Comment:
It's definitely not a significant perf improvement. 

Just to make sure I understand what you mean by "double specify sources", are you thinking of a scenario where a customer adds nuget.org as an audit source, but they also want vulnerability info from one of their package sources, so now they're going to have to specify that package source both in package sources and audit sources?  If you mean something different, then it's not intuitive to me (and frankly this interpretation of your comment feels like a guess)

On the other hand, consider a scenario where a project removes a PackageReference, or switches between branches (both of which have previously been restored, so global packages folder is populated). There's a good chance that NuGet won't need to query any package sources during the graph resolution. If NuGetAudit contacts package sources for the vulnerability info, then that might trigger interactive credential acquisition. So, the ability to not use all package sources as audit sources can also be considered a UX improvement.

But yeah, it's a UX change. As mentioned during the meeting, I regret the initial NuGetAudit implementation. I don't like package sources being overloaded for multiple purposes. I think it would be less confusing to customers if audit sources were only explicitly defined. If I could have gone back in time, I would have made this the first implementation, and if we got feedback then change it to what we currently have. But there are many examples of breaking changes making CX improvements difficult or impossible 😢

Reply:
> Just to make sure I understand what you mean by "double specify sources", are you thinking of a scenario where a customer adds nuget.org as an audit source, but they also want vulnerability info from one of their package sources, so now they're going to have to specify that package source both in package sources and audit sources

Yes, you'd basically need to specify nuget.org in both sources & audit sources.

Now I see that your spec says, when there's at least one. 
Not really sure how I feel about the whole thing tbh, it's a bit confusing, it doesn't affect people who have nuget.org in their sources already, so that concern is addressed.